### PR TITLE
mgr/rook: disallow_untyped_defs = True

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/rook/rook_cluster.py [hunk 1]
```diff
@@ -725,7 +744,7 @@ def _execute_blight_job(self, ident_fault: str, on: bool, loc: orchestrator.Devi
                 deleted = not api_response.items
                 if retries > 5:
                     sleep(0.1)
-                ++retries
+                retries = retries + 1
```
Comment (tchaikov) [inline src/pybind/mgr/rook/rook_cluster.py:747]:
```suggestion
                retries += 1
```

also, might want to move this change into a separate commit?

File: src/pybind/mgr/rook/rook_cluster.py [hunk 2]
```diff
@@ -74,17 +80,17 @@ class ApplyException(orchestrator.OrchestratorError):
     """
 
 
-def threaded(f):
-    def wrapper(*args, **kwargs):
+def threaded(f: FuncT) -> FuncT:
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
```
Comment (tchaikov) [inline src/pybind/mgr/rook/rook_cluster.py:84]: the `wrapper()` returns a `threading.Thread`, i think.

File: src/pybind/mgr/rook/rook_cluster.py [hunk 3]
```diff
@@ -74,17 +80,17 @@ class ApplyException(orchestrator.OrchestratorError):
...
         t = threading.Thread(target=f, args=args, kwargs=kwargs)
         t.start()
         return t
 
-    return wrapper
+    return cast(FuncT, wrapper)
```
Comment (tchaikov) [inline src/pybind/mgr/rook/rook_cluster.py:89]: i disagree. the `wrapper()` actually changes the return type of the wrapped function to `threading.Thread`.

File: src/pybind/mgr/rook/rook_cluster.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: done

File: src/pybind/mgr/rook/rook_cluster.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py:84]: that would work with something like 

```diff
diff --git a/src/pybind/mgr/rook/rook_cluster.py b/src/pybind/mgr/rook/rook_cluster.py
index f1f17afd50..ead55f60cc 100644
--- a/src/pybind/mgr/rook/rook_cluster.py
+++ b/src/pybind/mgr/rook/rook_cluster.py
@@ -80,13 +80,13 @@ class ApplyException(orchestrator.OrchestratorError):
     """
 
 
-def threaded(f: FuncT) -> FuncT:
-    def wrapper(*args: Any, **kwargs: Any) -> Any:
+def threaded(f: Callable[..., None]) -> Callable[..., threading.Thread]:
+    def wrapper(*args: Any, **kwargs: Any) -> threading.Thread:
         t = threading.Thread(target=f, args=args, kwargs=kwargs)
         t.start()
         return t
 
-    return cast(FuncT, wrapper)
+    return cast(Callable[..., threading.Thread], wrapper)
 
 
 class KubernetesResource(Generic[T]):
```

but this looses the arguments: See https://stackoverflow.com/questions/65936408/typing-function-when-decorator-change-generic-return-type

File: src/pybind/mgr/rook/rook_cluster.py (hunk 3 above)
Comment (tchaikov) [reply src/pybind/mgr/rook/rook_cluster.py:89]: i think cannot type this decorator properly before https://github.com/python/mypy/issues/3157 is fixed.

File: src/pybind/mgr/rook/rook_cluster.py (hunk 2 above)
Comment (tchaikov) [reply src/pybind/mgr/rook/rook_cluster.py:84]: can we have a yet-another-wrapper around `threaded()` to ensure the return type is correct typed, and keep the hack in the wrapped wrapper.

Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: After seeing the answer form @freundTech I went with the `Callable[...` solution.

**Comment by sebastian-philipp** [comment]
doc bulid:
```
Exception occurred:
  File "/home/jenkins-build/build/workspace/ceph-pr-docs/src/pybind/mgr/rook/rook_cluster.py", line 30, in <module>
    from rook.rook_client._helper import CrdClass
ModuleNotFoundError: No module named 'rook.rook_client._helper'; 'rook.rook_client' is not a package
```

**Review by tchaikov — APPROVED** [review]

**Review by tchaikov — COMMENTED** [review]
as long as the doc build is happy.

**Comment by tchaikov** [comment]
@sebastian-philipp probably you could apply follow patch
```diff
diff --git a/doc/_ext/ceph_commands.py b/doc/_ext/ceph_commands.py
index 58daba8f1de..f1d7ee606fe 100644
--- a/doc/_ext/ceph_commands.py
+++ b/doc/_ext/ceph_commands.py
@@ -254,6 +254,7 @@ class CephMgrCommands(Directive):
                          'jsonpatch',
                          'rook.rook_client',
                          'rook.rook_client.ceph',
+                         'rook.rook_client._helper',
                          'cherrypy=3.2.3']

         # make restful happy
```

**Comment by tchaikov** [comment]
```
Exception occurred:
  File "/home/jenkins-build/build/workspace/ceph-pr-docs/src/pybind/mgr/rook/rook_cluster.py", line 191, in RookCluster
    def __init__(self, coreV1_api: client.CoreV1Api, batchV1_api: client.BatchV1Api, rook_env: RookEnv):
NameError: name 'client' is not defined
```

then we need to mock `client` ...

**Comment by sebastian-philipp** [comment]
> ```
> Exception occurred:
>   File "/home/jenkins-build/build/workspace/ceph-pr-docs/src/pybind/mgr/rook/rook_cluster.py", line 191, in RookCluster
>     def __init__(self, coreV1_api: client.CoreV1Api, batchV1_api: client.BatchV1Api, rook_env: RookEnv):
> NameError: name 'client' is not defined
> ```
> 
> then we need to mock `client` ...

this would void this runtime check:

https://github.com/ceph/ceph/blob/39fd806cd45e991deac49c8a938dd6381306facf/src/pybind/mgr/rook/module.py#L114-L115

The reason for this runtime check is that some (especially dev) environments don't have this installed. I'll have a look

File: src/pybind/mgr/rook/rook_cluster.py [hunk 4]
```diff
@@ -183,29 +188,31 @@ def _watch(self, res_ver):
 
 
 class RookCluster(object):
-    def __init__(self, coreV1_api, batchV1_api, rook_env):
+    # import of client.CoreV1Api must be optional at import time.
+    # Instead allow mgr/rook to be imported anyway.
+    def __init__(self, coreV1_api: 'client.CoreV1Api', batchV1_api: 'client.BatchV1Api', rook_env: RookEnv):
```
Comment (tchaikov) [inline src/pybind/mgr/rook/rook_cluster.py:193]:
```suggestion
    def __init__(self, coreV1_api: 'client.CoreV1Api', batchV1_api: 'client.BatchV1Api', rook_env: 'RookEnv'):
```

Comment (tchaikov) [reply src/pybind/mgr/rook/rook_cluster.py:193]: or just define dummy classes for these types if `.rook_client._helper` and `rook.module` are not available?

**Review by tchaikov — APPROVED** [review]

File: src/pybind/mgr/rook/rook_cluster.py [hunk 5]
```diff
@@ -372,14 +381,13 @@ def remove_pods(self, names):
                     self.rook_env.namespace,
                     body=client.V1DeleteOptions()
                 )
-                num += 1
-        return "Removed %d pods" % num
+        return [f'Removed Pod n' for n in names]
```
Comment (varshar16) [inline src/pybind/mgr/rook/rook_cluster.py:384]: Let it just return pod names ```[f'n' for n in names]```

**Review by varshar16 — APPROVED** [review]

File: src/pybind/mgr/rook/rook_cluster.py (hunk 5 above)
Comment (tchaikov) [reply src/pybind/mgr/rook/rook_cluster.py:384]: do you mean to say 
```suggestion
        return [n for n in names]
```
?

Comment (varshar16) [reply src/pybind/mgr/rook/rook_cluster.py:384]: > do you mean to say
> 
> ?

Yes

Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py:384]: this is the corresponding string in cephadm *per daemon*

https://github.com/ceph/ceph/blob/ae167bd496d7dc0a72a24d7c393c32c5420c6bc2/src/pybind/mgr/cephadm/serve.py#L858

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-01-29_12:57:23-rados-wip-kefu-testing-2021-01-29-1713-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-30_16:51:13-rados-wip-kefu-testing-2021-01-30-2256-distro-basic-smithi/

**Comment by tchaikov** [comment]
rebased against master.
